### PR TITLE
Add ROCm support (AMDGPU)

### PR DESCRIPTION
**Pull request by luraess**
Add AMDGPU.ROCArray support to the MPI machinery to allow for ROCm-aware MPI device pointer exchange, as suggested by @simonbyrne.

**:warning: Note that this PR should replace and close #547 (without merging the latter) as suggested in https://github.com/JuliaParallel/MPI.jl/pull/547#issuecomment-1101671327** 

Requires AMDGPU v0.3.5 to work smoothly. (Thanks to @utkinis for helping out as well).

Most MPI.jl tests pass exporting `JULIA_MPI_TEST_ARRAYTYPE=ROCArray` with exception of:
- `test_reduce.jl`
- `test_subarray.jl`
- `test_threads.jl`

Failing test are currently commented out and preceded by `# DEBUG` comment.

The `dev` doc should also include references to the latest changes.

### To-dos
- to pass the tests, #564 is currently needed (issue #561) and there seems to still be an issue running the MPI tests outside of the MPI.jl project. Would require merging (#564) before this PR.
- selecting system MPI is still not super stable (#570)
- currently missing the analogy of [`CUDA.precompile_runtime()`](https://github.com/luraess/MPI.jl/blob/ef153a261ea73b48f6757b47065df8a485890880/test/runtests.jl#L10) for AMDGPU. The feature would first need to be added to AMDGPU.jl (note that all works fine without for now)
- one may want to add the CUDA-aware MPI + multi-GPU MWE to `usage.md` in the doc as well?
- add the correct definition to `.buildkite/pipeline.yml`(I am waiting to hear back from the sysadmin that built the ROCm-aware OpenMPI on the test server regarding the config)

### CODE REVIEW

File: test/test_threads.jl [hunk 1]
```diff
@@ -4,39 +4,43 @@ using MPI
...
     ArrayType = CUDA.CuArray
+elseif get(ENV,"JULIA_MPI_TEST_ARRAYTYPE","") == "ROCArray"
+    import AMDGPU
+    ArrayType = AMDGPU.ROCArray
 else
     ArrayType = Array
 end
 
-provided = MPI.Init(threadlevel=:multiple)
+# DEBUG: currently failing
```
Comment (vchuravy) [inline test/test_threads.jl:14]: Any intuition why these are failing? Is your local MPI not build with multiple thread support?

File: test/test_reduce.jl [hunk 2]
```diff
@@ -97,10 +100,11 @@ for T = [Int]
 
             # Allocating, Subarray
             recv_arr = MPI.Reduce(view(send_arr, 2:3), op, MPI.COMM_WORLD; root=root)
-            if isroot
-                @test recv_arr isa ArrayType{T}
-                @test recv_arr == sz .* view(send_arr, 2:3)
-            end
+            # DEBUG: currently failing with ROCArray
+            # if isroot
```
[the remark this answers is not in the document]
Comment (luraess) [reply test/test_reduce.jl:104]: Failing then crashing.

File: test/test_threads.jl (hunk 1 above)
Comment (luraess) [reply test/test_threads.jl:14]: No intuition on that. Maybe no multiple thread support indeed. How could I check this?

Comment (vchuravy) [reply test/test_threads.jl:14]: If you are using OpenMPI use ompi_info

Comment (luraess) [reply test/test_threads.jl:14]:
```
[lraess@ault20 dev]$ ompi_info | grep "Thread support"
          Thread support: posix (MPI_THREAD_MULTIPLE: yes, OPAL support: yes, OMPI progress: no, ORTE progress: yes, Event lib: yes)
```

File: test/test_basic.jl [hunk 3]
```diff
@@ -16,7 +19,8 @@ MPI.Init()
 
 @test MPI.has_cuda() isa Bool
 
-if ArrayType != Array
+# DEBUG: a cleaner apporach may be designed
+if ArrayType != Array && ArrayType != AMDGPU.ROCArray
```
[the remark this answers is not in the document]
Comment (luraess) [reply test/test_basic.jl]: Good point.

**Comment by luraess** [comment]
Add ROCm test to Buildkite. 

Status (cc @vchuravy) - i.e. curent issues:
- Some unit tests fail in the GH pipeline related to AMDGPU because of `libhsa` not found ([see here](https://github.com/JuliaParallel/MPI.jl/runs/6254036631?check_suite_focus=true#step:6:42))
- The Buildkite test fail for both CUDA and ROCm; segfaults in CUDA and `The value of the MCA parameter "plm_rsh_agent" was set to a path that could not be found: plm_rsh_agent: ssh : rsh` in ROCm.
- Note that although the the passing Buildkite status (also in #master) is misleading as MPI build pass, but tests fail

**Comment by giordano** [comment]
> Some unit tests fail in the GH pipeline related to AMDGPU because of libhsa not found ([see here](https://github.com/JuliaParallel/MPI.jl/runs/6254036631?check_suite_focus=true#step:6:42))

It's the i686 jobs that are failing, because the library isn't available for that platform: https://github.com/JuliaPackaging/Yggdrasil/blob/b249dffd4df24711384874fda84ffd926f859af9/H/hsa_rocr/build_tarballs.jl#L36-L39

**Comment by luraess** [comment]
> because the library isn't available for that platform

Thanks for insights. What workaround could you suggest?

**Comment by giordano** [comment]
I have no idea whether the library can be built for that platform, but the name `libhsa-runtime64` doesn't suggest so.  If it can be built, you should try and add `Platform("i686", "linux")` on Yggdrasil, but if it can't you'd simply skip the test here.

**Comment by vchuravy** [comment]
@jpsamaroo can we make AMDGPU loadable everywhere?

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -161,3 +161,44 @@ steps:
...
+      rocm: "*" # todo fix ROCM version
+    if: build.message !~ /\[skip tests\]/
+    timeout_in_minutes: 60
+    env:
+      JULIA_MPI_TEST_ARRAYTYPE: ROCArray
+      JULIA_MPI_TEST_NPROCS: 2
+      JULIA_MPI_PATH: "${BUILDKITE_BUILD_CHECKOUT_PATH}/openmpi"
+      OMPI_ALLOW_RUN_AS_ROOT: 1
+      OMPI_ALLOW_RUN_AS_ROOT_CONFIRM: 1
+      OMPI_MCA_btl_vader_single_copy_mechanism: 'none' # https://github.com/open-mpi/ompi/issues/4948
```
Comment (vchuravy) [inline .buildkite/pipeline.yml:181]:
```suggestion
      OMPI_MCA_btl_vader_single_copy_mechanism: 'none' # https://github.com/open-mpi/ompi/issues/4948
      OMPI_MCA_plm_rsh_agent: 'sh' # the container doesn't have ssh installed, but we don't need it
```

**Comment by vchuravy** [comment]
> is misleading as MPI build pass, but tests fail

Yeah that is intentional while we figure out the bugs.

**Comment by luraess** [comment]
The `plm_rsh_agent` changes are committed. Now there are new errors (related to UCX?). I cannot interpret those that well tbh.

**Comment by jpsamaroo** [comment]
Working on a fix in https://github.com/JuliaGPU/AMDGPU.jl/pull/229

**Comment by luraess** [comment]
> Working on a fix in [JuliaGPU/AMDGPU.jl#229](https://github.com/JuliaGPU/AMDGPU.jl/pull/229)

Committed the latest changes from main into the branch used for this PR which re-triggered CI. Now the only still failing tests relate to `mpitrampoline-system-apt` and Buildkite (possibly UCX issues there).

File: test/test_reduce.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (vchuravy) [reply test/test_reduce.jl:104]: Uncomment this please.This is likely similar to #563 in that we need to make sure the GPU is done before we call MPI.

File: test/test_subarray.jl [hunk 5]
```diff
@@ -34,7 +37,8 @@ src  = mod(rank-1, comm_size)
 
     MPI.Waitall([req_send, req_recv])
 
-    @test X[3:4,1] == Y
+    # DEBUG: currently failing with ROCArray
+    # @test X[3:4,1] == Y
```
Comment (vchuravy) [inline test/test_subarray.jl:41]: Same here. Either mark with `@test_broken` or uncomment

**Review by vchuravy — CHANGES_REQUESTED** [review]
None of the test should be uncommented. You can either mark them with `@test_broken` (only for AMD) or just let them fail for now. I think then we can merge this and work on CI on master

**Comment by giordano** [comment]
@eschnett can you please take a look at the [mpitrampoline failures](https://github.com/JuliaParallel/MPI.jl/runs/6430518478?check_suite_focus=true#step:9:272)?
```
MPItrampoline: MPI ABI version mismatch:
This version of MPItrampoline requires MPI ABI version 2.7.0, but the loaded MPIwrapper only provides MPI ABI version 2.2.0.
This is MPItrampoline version 4.0.2.
You loaded MPIwrapper version 2.3.3 from file "/usr/local/lib/libmpiwrapper.so".
```

**Comment by eschnett** [comment]
@giordano This look as if you are using an outdated version of `MPIwrapper`. If you built and installed MPIwrapper yourself, then you will need to install a newer version (current in 2.8.1).

**Comment by giordano** [comment]
So we need to change https://github.com/JuliaParallel/MPI.jl/blob/5f25dcc616afe64b1e8a26710a94fa614e65af95/.github/workflows/UnitTests.yml#L404-L405 to 2.8.1?

**Comment by luraess** [comment]
Shall one update `MPIWrapper` for all MPI trampoline based Unit tests?

**Comment by luraess** [comment]
Also, now `test_io_shared` fails on `test-default (windows-latest, 1.6, x64)`

**Comment by giordano** [comment]
I've taken out fixing CI to #596, so that we can have working CI without having to wait for this PR to be reviewed (although that should hopefully not take much longer), and also because that's a logically unrelated change anyway.

**Comment by luraess** [comment]
Can this PR be (soon?) merged @vchuravy (cc @simonbyrne)?

**Comment by simonbyrne** [comment]
@vchuravy can we fix the buildkite CI as well?

**Comment by luraess** [comment]
> fix the buildkite CI

ROCm OpenMPI fails because of the MPI build (UCX errors), CUDA MPI errors on other stuff.

**Comment by simonbyrne** [comment]
the rocm queue doesn't seem to be working at all at the moment:
https://buildkite.com/julialang/mpi-dot-jl/builds/737

**Comment by simonbyrne** [comment]
@jpsamaroo is this something you can help with?

**Comment by luraess** [comment]
> the rocm queue doesn't seem to be working

Indeed, maybe the machines are "down". It used to build successfully, but fail with and UCX error in the tests.

**Comment by simonbyrne** [comment]
I got CUDA tests working again in #607, but disabled the ROCm builds: we can re-enable that once the machines are back online.

I also reconfigured the tests: you now should only need to modify `test/common.jl` to add support for ROCm.

**Comment by luraess** [comment]
> I got CUDA tests working again in #607, but disabled the ROCm builds: we can re-enable that once the machines are back online.

Thanks @simonbyrne. AMDGPU server should be back up soon. I suppose we can then add back a `ROCm` group by analogy of the new CUDA setup. I guess ROCm UCX would need a fix also.

> I also reconfigured the tests: you now should only need to modify `test/common.jl` to add support for ROCm.

I will resolve conflicts including latest changes to this PR.

**Comment by luraess** [comment]
> I also reconfigured the tests: you now should only need to modify `test/common.jl` to add support for ROCm.

@simonbyrne: I included the changes to tests to account for AMDGPU. Note that there is an issue wrt `synchronize()` as added in #563 because AMDGPU does not sync using `wait( )` to wrap around the `@roc` kernel call.

**Comment by simonbyrne** [comment]
Do you want [`hipDeviceSynchronize`](https://github.com/JuliaGPU/AMDGPU.jl/blob/43124078c1c5b6359967877d1f1dab0ba4acdf2b/src/hip/libhip.jl#L9)?

**Comment by vchuravy** [comment]
No that is the wrong device synchronize. @jpsamaroo we probably want https://github.com/JuliaGPU/KernelAbstractions.jl/blob/5ef9c191fb3030870b83e3f10c76da92e15c45a4/lib/ROCKernels/src/ROCKernels.jl#L84 + hip?

**Comment by luraess** [comment]
I understood we don't want to depend on HIP, but directly on HSA. And I don't know if the current CPU `wait()` can be implemented in a syntax that does not need the wrapping around the `@roc` call.

**Comment by vchuravy** [comment]
@jpsamaroo can you open a PR to AMDGPU to provide a less wordy vairant of `AMDGPU.barrier_and!(queue, AMDGPU.active_kernels(AMDGPU.get_default_queue()))`? and we will probably want to have that include the hip call as well.

**Comment by luraess** [comment]
> @jpsamaroo can you open a PR to AMDGPU to provide a less wordy vairant of `AMDGPU.barrier_and!(queue, AMDGPU.active_kernels(AMDGPU.get_default_queue()))`? and we will probably want to have that include the hip call as well.

Now there is `wait(@roc ...)`. If the proposed variant would act as `synchronize()` in CUDA then it could be quite a breaking? change to the AMDGPU API which may have some broader impact. It could be good, but let's not forget consistency...

**Comment by simonbyrne** [comment]
I've added AMD back to buildkite on this PR, but it doesn't look like it is set up correctly.

**Comment by simonbyrne** [comment]
Test failure could possibly be https://github.com/openucx/ucx/issues/5485

@luraess were you able to run the tests locally?

**Comment by luraess** [comment]
@simonbyrne The tests are running fine locally for me. I asked the sysadmin if they could share some info about the config of the ROCm-aware OpenMPI build (it's OpenMPI v4.0.6rc4 with ROCm 4.3). Also, one needs Julia 1.7.

**Comment by luraess** [comment]
@simonbyrne The tests are running locally for me (with exception for spawn which error due to allocation issues running things through `srun`). I asked the sysadmin if they could share some info about the config of the ROCm-aware OpenMPI build (it's OpenMPI v4.0.6rc4 with ROCm 4.3). Also, one needs Julia 1.7.

Regarding the test failure related to ipc, I had similar issue when running on non ROCm-aware MPI and found that adding `export PMIX_MCA_psec=native` helped.

File: test/common.jl [hunk 6]
```diff
@@ -5,6 +5,10 @@ if get(ENV,"JULIA_MPI_TEST_ARRAYTYPE","") == "CuArray"
     import CUDA
     ArrayType = CUDA.CuArray
     synchronize() = CUDA.synchronize()
+elseif get(ENV,"JULIA_MPI_TEST_ARRAYTYPE","") == "ROCArray"
+    import AMDGPU
+    ArrayType = AMDGPU.ROCArray
+    synchronize() = nothing
```
Comment (simonbyrne) [inline test/common.jl:11]: Okay, now the synchronization appears to be the issue: what can I use for this?

[the remark this answers is not in the document]
Comment (vchuravy) [reply test/common.jl:11]: I think the above is about right? I don't like it...

[the remark this answers is not in the document]
Comment (simonbyrne) [reply test/common.jl:11]: What is `queue`?

[the remark this answers is not in the document]
Comment (luraess) [reply test/common.jl:11]: It's like the analogy of the CUDA stream.

[the remark this answers is not in the document]
Comment (luraess) [reply test/common.jl:11]: > I think the above is about right? I don't like it...

Maybe @jpsamaroo has some thoughts?

File: test/common.jl [hunk 7]
```diff
@@ -5,6 +5,16 @@ if get(ENV,"JULIA_MPI_TEST_ARRAYTYPE","") == "CuArray"
...
     ArrayType = CUDA.CuArray
     synchronize() = CUDA.synchronize()
+elseif get(ENV,"JULIA_MPI_TEST_ARRAYTYPE","") == "ROCArray"
+    import AMDGPU
+    ArrayType = AMDGPU.ROCArray
+    function synchronize()
+        # TODO: AMDGPU synchronization story is complicated. HSA does not provide a consistent notion of global queues. We need a mechanism for all GPUArrays.jl provided kernels to be synchronized.
+        queue = AMDGPU.get_default_queue()
+        barrier = AMDGPU.barrier_and!(queue, AMDGPU.active_kernels(queue))
+        AMDGPU.hipDeviceSynchronize() # Sync all HIP kernels e.g. BLAS. N.B. this is blocking Julia progress
```
Comment (simonbyrne) [inline test/common.jl:15]:
```suggestion
        AMDGPU.HIP.hipDeviceSynchronize() # Sync all HIP kernels e.g. BLAS. N.B. this is blocking Julia progress
```

File: test/common.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpsamaroo) [reply test/common.jl]: It's `AMDGPU.HIP.hipDeviceSynchronize()`. HIP and HSA functions are behind their respective modules.

**Comment by luraess** [comment]
> what UCX version is your OpenMPI install using?

@vchuravy Here the output of `ompi_info` (see [ompi_info.txt](https://github.com/JuliaParallel/MPI.jl/files/8827319/ompi_info.txt)) and `ucx_info` (see [ucx_info.txt](https://github.com/JuliaParallel/MPI.jl/files/8827325/ucx_info.txt))  on ault.

**Comment by simonbyrne** [comment]
So i think it is failing on this line:
https://github.com/JuliaParallel/MPI.jl/blob/5fd4180e4a3aa24cf97f6c66ef5eb9069de8e552/test/test_reduce.jl#L93-L97
How are views handled for ROCArrays?

File: src/rocm.jl [hunk 8]
```diff
@@ -1,11 +1,11 @@
 import .AMDGPU
 
 function Base.cconvert(::Type{MPIPtr}, A::AMDGPU.ROCArray{T}) where T
-    Base.cconvert(Ptr{T}, A.buf.ptr) # returns DeviceBuffer
+    A
```
Comment (luraess) [inline src/rocm.jl:4]: What was wrong here @simonbyrne ?

**Comment by luraess** [comment]
Cool, now the tests pass:

**Comment by jpsamaroo** [comment]
What's up with the failing 1.7 job?

**Comment by simonbyrne** [comment]
> ERROR: LoadError: LoadError: No GPU agents detected!
Please consider rebuilding AMDGPU

I've been seeing it intermittently. It usually goes away with on a restart though. Is it an issue with the runner?

File: src/rocm.jl (hunk 8 above)
Comment (simonbyrne) [reply src/rocm.jl:4]: `A.buf.ptr` is a raw pointer, so the object could be cleaned up by GC. It also loses the `offset` argument required for views.

**Comment by jpsamaroo** [comment]
Ahh, yes, this was an off-by-one bug in the amdgpuci.8 runner config; it should we working correctly now!

**Review by simonbyrne — APPROVED** [review]

File: src/rocm.jl (hunk 8 above)
Comment (luraess) [reply src/rocm.jl:4]: Thanks! TIL

**Comment by vchuravy** [comment]
Fantastic work all around!

**Pull request merged**